[RequirementMachine] Implement inference and minimization of same-shape requirements.

hborla opened the pull request:
A "same-shape" requirement states that two type parameter packs have the same structure, including having the same length. Same-shape requirements cannot be written explicitly; they can only be inferred from same-type requirements between parameter packs and pack expansions that contain multiple parameter packs:

```swift
// '(T, U)...' implies a same-shape requirement between T and U.
func inferSameShape<@_typeSequence T, @_typeSequence U>(ts t: T..., us u: U...) where ((T, U)...): Any
```

Same-shape requirements between dependent member types are desugared to equate the root type parameter shapes:

```swift
protocol P {
  associatedtype A
}

// '(T.A, U.A)...' implies a same-shape requirement between T and U.
func inferSameShape<@_typeSequence T, @_typeSequence U>(ts t: T..., us u: U...) where ((T.A, U.A)...): Any
```

A same-shape requirement is a rewrite rule:

```
T.[shape] => U.[shape]
```

`[shape]` is a new symbol that can be appended to generic parameter terms. In the future, we can allow concrete shape requirements using a concrete shape property symbol, e.g.

```
T.[shape].[concreteShape: PackType(Int, U...)] => T.[shape]
```

hborla committed:
[RequirementMachine] Add same-length requirement inference for pack
expansion types.

hborla committed:
[RequirementMachine] Add a new symbol kind for the shape of a parameter
pack.

hborla committed:
[RequirementMachine] Map same-shape requriements to rewrite rules.

A same-shape requirement 'length(T...) == length(U...)' becomes a rewrite
rule 'T.[shape] => U.[shape]'. Reduced shape rules will drop the [shape]
term from each side of the rule, and create a same-shape requirement
between the two type parameter packs.

hborla committed:
[RequirementMachine] Generalize connected components to process same
shape requirements.

hborla committed:
[RequirementMachine] Add tests for shape requirements made redundant by
same-type requirements.

slavapestov commented on lib/AST/RequirementMachine/Symbol.cpp:
```diff
@@ -295,6 +306,30 @@ Symbol Symbol::forGenericParam(GenericTypeParamType *param,
+  id.AddInteger(unsigned(Kind::Shape));
+
+  void *insertPos = nullptr;
+  if (auto *symbol = ctx.Symbols.FindNodeOrInsertPos(id, insertPos))
```
Since this is a singleton, we should store it in a ctx.TheShapeSymbol variable or something, instead of going through the folding set?

hborla replied:
Are you suggesting moving the code that creates the symbol storage into the `RewriteContext` constructor to assign it into a `RewriteContext::TheShapeSymbol` field, or something else? Only thing is that the `Storage` type is currently only visible in Symbol.cpp. I'm happy to move code around, just want to understand what you're thinking

slavapestov replied:
Yeah, I'm thinking adding a `Symbol::Storage *` to RewriteContext. Then instead of calling ctx.Symbols.FindNodeOrInsertPos() on ctx.Symbols you just check if ctx.TheShapeSymbol is non-null

hborla replied:
Oh I see, that's easier than constructing a `Storage` in RewriteContext, thanks!

hborla committed:
[RequirementMachine] Add a generic signature query that returns the
reduced shape of a given type parameter pack.

hborla committed:
[Requirement] Rename RequirementKind::SameCount to SameShape.

slavapestov commented on lib/AST/RequirementMachine/GenericSignatureQueries.cpp:
```diff
@@ -696,6 +696,26 @@ RequirementMachine::lookupNestedType(Type depType, Identifier name) const {
+  if (!type->isTypeSequenceParameter())
+    return Type();
+
+  auto term = Context.getMutableTermForType(type->getCanonicalType(),
```
You need to get the root generic parameter here before converting to a term

hborla committed:
[RequirementMachine] Simplify same-shape requirement inference from pack
expansion types by equating the pack expansion count/shape type with each
shape of the referenced packs in the expansion.

hborla commented on lib/AST/RequirementMachine/RequirementLowering.cpp:
```diff
@@ -474,6 +493,23 @@ struct InferRequirementsWalker : public TypeWalker {
+        auto first = packReferences.begin();
+        auto second = first + 1;
+        while (second != packReferences.end()) {
+          Requirement req(RequirementKind::SameCount, *first++, *second++);
```
Good idea!

hborla committed:
[NFC][RequirementMachine] Minor syntax change for MutableTerm initialization.

hborla committed:
[RequirementMachine] Add verification for rules involving shape symbols.

hborla committed:
[RequirementMachine] Store the singleton 'Storage' instance for shape
symbols in RewriteContext.

hborla committed:
[RequirementMachine] Add a generic signature query that determines whether
two type parameter packs have the same shape.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test

hborla commented:
Windows test failure is unrelated

```
Failed Tests (1):
  Swift(windows-x86_64) :: Interop/Cxx/class/mutability-annotations-typechecker.swift
```
